docs/blog/2021-12-01-csp.md

```diff
@@ -0,0 +1,208 @@
+
+As you can see, these are nuanced details that are easy to miss.
+
+### Other vulnerabilities
+
+These are just a few of the potential ways to bypass the CSP's protections. The
+list [goes on](https://book.hacktricks.xyz/pentesting-web/content-security-policy-csp-bypass).
+
+### Mitigations
+
+For large sites, there may be just too many gaps in the armor to treat the CSP as a functional line of defense, and
+many big sites have disabled it entirely.
+[This paper](https://research.google/pubs/pub45542/) found that of those that use it **over 94% of sites have
+ineffective CSPs!**
+
+Long story short, there's a good chance your CSP isn't doing much to protect your site from XSS.
+
+Nonce-based CSP strategies can help mitigate the above attacks, where the server adds a random nonce attribute to every
+script tag and defines in the CSP. Embedding a script tag smart enough to clone the nonce isn't possible without
+JavaScript.
+
+Unfortunately this strategy is the most work to implement and is rarely used.  Getting it to work with CDNs and caching can be a challenge.
+
+```text title=CSP
+script-src 'nonce-randomString';
+```
+
+```html title="Script Tag"
+<script nonce="randomString" ...
+```
+
+Defining a nonce and a domain `script-src` together means that _either_ the domain or the nonce can be used, which seems
+like a gotcha that could catch many users.
+
+## How LunaSec and other tools use CSP effectively
+
+LunaSec is an Open Source security product we wrote, designed to help normal web apps handle sensitive data. The
+main idea with LunaSec is to be a secure "sidecar" for every part of a full-stack web app.
+
+The core of its frontend security is that it embeds sensitive form
+elements [inside an iFrame](https://www.lunasec.io/docs/pages/how-it-works/secure-components/) on a different domain
+than the main site. This is very similar to how payment processors like Stripe embed an iFrame to collect payment
+information, just more generic.
+
+The above problems with CSP are a great demonstration of why we've worked to try to isolate and secure a small portion
+of our user's pages, instead of trying to completely secure the main sites themselves. Big production web apps are just
+too large of an attack surface for the average company to fully audit for vulnerabilities.
+
+LunaSec and frameworks like it embed a tiny, stripped down, domain specific webpage that does one thing, totally focused
+on security. Because it's on a different domain, (in LunaSec's case, usually a subdomain you own) the main site can be
+vulnerable to the above issues without exposing the sensitive data to XSS.
+
+Here is an example CSP header that loads with one of LunaSec's iFrames: 
+```text
+Content-Security-Policy: 
+style-src 'unsafe-inline' 'nonce-r54BDFpZ8Nkb9IffY7JwJA';
+script-src 'nonce-r54BDFpZ8Nkb9IffY7JwJA';
+object-src 'none';
+frame-ancestors https://your-site.com;
+base-uri 'none';
+report-uri http://localhost:5004;
```

Comment:
we probably want to change this to `http://lunasec-secure-backend.your-site.com/cspreport`. `http://localhost:5004` is just a placeholder for now.